File: web/src/lib/earnings.ts
/**
 * Client-side earnings preview, matching the backend and SQL contract.
 *
 * hours = (tasks_attempter * aht_attempter_minutes
 *        + tasks_reviewer * aht_reviewer_minutes) / 60
 * earnings = hours * USD hourly rate
 *
 * Historical logs use their immutable snapshot fields only.
 */

export function computeHours(
  tasksAttempter: number,
  tasksReviewer: number,
  ahtAttempterMinutes: number,
  ahtReviewerMinutes: number,
): number {
  const totalMinutes =
    Number(tasksAttempter) * Number(ahtAttempterMinutes) +
    Number(tasksReviewer) * Number(ahtReviewerMinutes);
  return totalMinutes / 60;
}

export function computeEarnings(hours: number, hourlyRate: number): number {
  return Number(hours) * Number(hourlyRate);
}

/** Resolve partially backfilled USD earnings without hiding a positive value. */
export function resolveEarningsUsd(
  calculatedEarningsUsd: unknown,
  calculatedEarnings: unknown,
): number {
  const usd = Number(calculatedEarningsUsd);
  const fallback = Number(calculatedEarnings);

  if (Number.isFinite(usd) && usd === 0 && Number.isFinite(fallback) && fallback > 0) {
    return fallback;
  }
  if (Number.isFinite(usd)) return usd;
  if (Number.isFinite(fallback)) return fallback;
  return 0;
}

export function ratePerTask(ahtMinutes: number, hourlyRate: number): number {
  return (Number(ahtMinutes) / 60) * Number(hourlyRate);
}

export function computePreview(
  tasksAttempter: number,
  tasksReviewer: number,
  ahtAttempterMinutes: number,
  ahtReviewerMinutes: number,
  hourlyRate: number,
): {
  hours: number;
  earnings: number;
  rate_per_task_attempter: number;
  rate_per_task_reviewer: number;
} {
  const hours = computeHours(
    tasksAttempter,
    tasksReviewer,
    ahtAttempterMinutes,
    ahtReviewerMinutes,
  );
  return {
    hours,
    earnings: computeEarnings(hours, hourlyRate),
    rate_per_task_attempter: ratePerTask(ahtAttempterMinutes, hourlyRate),
    rate_per_task_reviewer: ratePerTask(ahtReviewerMinutes, hourlyRate),
  };
}

/** Hours from a historical log using immutable snapshot fields only. */
export function hoursFromLog(log: {
  tasks_attempter: number;
  tasks_reviewer: number;
  snapshot_aht_attempter: number;
  snapshot_aht_reviewer: number;
}): number {
  return computeHours(
    log.tasks_attempter,
    log.tasks_reviewer,
    log.snapshot_aht_attempter,
    log.snapshot_aht_reviewer,
  );
}

// Keep the existing earnings module as the import surface while the formatters live centrally.
export {
  formatAhtMinutes,
  formatCurrency,
  formatDate,
  formatHours,
  formatNumber,
} from "@/lib/formatters";

File: backend/app/services/earnings.py
"""
Pure earnings math for AnnotatePay.

Authoritative formula (document & enforce everywhere):

    hours = (tasks_attempter * aht_attempter_minutes
             + tasks_reviewer * aht_reviewer_minutes) / 60
    earnings = hours * hourly_rate
    rate_per_task = (aht_minutes / 60) * hourly_rate

Notes:
- AHT values are in MINUTES.
- Zero tasks are allowed for either role.
- Historical snapshots are immutable: when aggregating/exporting logs,
  always use the AHT and hourly_rate stored on each log row — never
  "fix" past logs with current project defaults.
"""

from __future__ import annotations


def compute_hours(
    tasks_attempter: int | float,
    tasks_reviewer: int | float,
    aht_attempter_minutes: float,
    aht_reviewer_minutes: float,
) -> float:
    """Return total hours from task counts and AHT (minutes)."""
    total_minutes = (
        float(tasks_attempter) * float(aht_attempter_minutes)
        + float(tasks_reviewer) * float(aht_reviewer_minutes)
    )
    return total_minutes / 60.0


def compute_earnings(hours: float, hourly_rate: float) -> float:
    """Return earnings = hours * hourly_rate."""
    return float(hours) * float(hourly_rate)


def rate_per_task(aht_minutes: float, hourly_rate: float) -> float:
    """
    Effective $ per single task for a role.

        rate_per_task = (aht_minutes / 60) * hourly_rate
    """
    return (float(aht_minutes) / 60.0) * float(hourly_rate)


def compute_preview(
    tasks_attempter: int | float,
    tasks_reviewer: int | float,
    aht_attempter_minutes: float,
    aht_reviewer_minutes: float,
    hourly_rate: float,
) -> dict[str, float]:
    """Full preview payload; all monetary values are canonical USD."""
    hours = compute_hours(
        tasks_attempter,
        tasks_reviewer,
        aht_attempter_minutes,
        aht_reviewer_minutes,
    )
    earnings = compute_earnings(hours, hourly_rate)
    return {
        "hours": hours,
        "earnings": earnings,
        "rate_per_task_attempter": rate_per_task(aht_attempter_minutes, hourly_rate),
        "rate_per_task_reviewer": rate_per_task(aht_reviewer_minutes, hourly_rate),
    }


def hours_and_earnings_from_log(
    tasks_attempter: int | float,
    tasks_reviewer: int | float,
    aht_attempter_minutes: float,
    aht_reviewer_minutes: float,
    hourly_rate: float,
    stored_hours: float | None = None,
    stored_earnings: float | None = None,
    stored_earnings_usd: float | None = None,
    snapshot_available: bool | None = None,
) -> tuple[float, float]:
    """
    Resolve hours/earnings for a historical log row.

    Prefer recomputing from available snapshot AHT/rate fields so analytics
    and exports stay consistent with the formula. If a row has no snapshot
    fields, fall back to its stored hours and canonical USD earnings.

    ``snapshot_available`` is supplied by row coercion when the source shape
    is known. The inferred default preserves the historical pure-function
    behavior for callers that pass only numeric values.
    """
    if snapshot_available is None:
        snapshot_available = (
            float(aht_attempter_minutes) != 0
            or float(aht_reviewer_minutes) != 0
            or float(hourly_rate) != 0
        )

    if snapshot_available:
        hours = compute_hours(
            tasks_attempter,
            tasks_reviewer,
            aht_attempter_minutes,
            aht_reviewer_minutes,
        )
        earnings = compute_earnings(hours, hourly_rate)
        return hours, earnings

    hours = float(stored_hours) if stored_hours is not None else 0.0
    stored_usd = (
        stored_earnings_usd
        if stored_earnings_usd is not None
        else stored_earnings
    )
    earnings = float(stored_usd) if stored_usd is not None else 0.0
    return hours, earnings
